File: C/stream.c
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <time.h>

#define BUFFER_SIZE 1024

int main() {
    clock_t start = clock();
    int infile = open("../input.txt", O_RDONLY);
    if (infile < 0) {
        perror("Error opening input file");
        return 1;
    }

    int outfile = open("stream_output.txt", O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (outfile < 0) {
        perror("Error creating output file");
        close(infile);
        return 1;
    }

    char buffer[BUFFER_SIZE];
    ssize_t bytesRead;
    while ((bytesRead = read(infile, buffer, BUFFER_SIZE)) > 0) {
        if (write(outfile, buffer, bytesRead) != bytesRead) {
            perror("Error writing to output file");
            close(infile);
            close(outfile);
            return 1;
        }
    }

    if (bytesRead < 0) {
        perror("Error reading input file");
    }

    close(infile);
    close(outfile);
    clock_t end = clock();
    double duration = ((double)(end - start)) / CLOCKS_PER_SEC;

    printf("Stream I/O Time taken: %f seconds\n", duration);
    return 0;
} 

File: Rust/src/stream.rs
use std::fs::File;
use std::io::{self, Read, Write};
use std::time::Instant;

fn main() -> io::Result<()> {
    let start = Instant::now();
    
    let mut input_file = File::open("../input.txt")?;
    let mut output_file = File::create("stream_output_rust.txt")?;
    
    let mut buffer = [0; 1]; // Reading one byte at a time
    loop {
        let bytes_read = input_file.read(&mut buffer)?;
        if bytes_read == 0 {
            break;
        }
        output_file.write_all(&buffer[..bytes_read])?;
    }
    
    let duration = start.elapsed();
    println!("Stream I/O Time taken: {:?}", duration);
    
    Ok(())
} 